rep-2007.rst

```diff
@@ -0,0 +1,246 @@
+This ambiguity may result in frustration on the part of the ROS 2 developer:
+
+* frequently having to refer back to documentation
+* possibly opaque error messages
+
+Facade
+^^^^^^
+
+"Facade" seems to be a more common English word than "masquerade".
+It also is commonly used as a design pattern in object oriented programming.
+However, the "Facade pattern" is typically used to simplify a complex interface [3]_, which is not the major feature being proposed here.
+
+It was thought to use "Facade" in the following form:
+
+.. code-block:: cpp
+
+   Facade<std::string>::instead_of<std_msgs::msg::String>
+
+
+Adapt
+^^^^^
+
+"Adapt" is certainly a common English word, and the "Adapter pattern" is a common design pattern for adjusting an interface [4]_, which matches well with the feature being suggested here.
+Also, using "Adapt" is consistent with the documentation of a similar feature in ROS 1 (i.e., "Adapting C++ Types" [5]_).
+
+"Adapt" also has the advantage of being a verb and of being related to the noun "Adapter".
+This flexiblity may make it easier for developers to discuss its use.
+
+"Adapt" could be used in the following syntax:
+
+.. code-block:: cpp
+
+   Adapt<std::string>::to<std_msgs::msg::String>
+
+Additional terms considered
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Here is a brief listing of additional terms that were considered and why they were not selected:
+
+:Convert: Passed in favor of "Adapt", which expresses a similar idea and has a common design pattern.
+
+:Decorate: Passed in favor of "Fascade", which seems to be more common.
+
+:Mask: Overloaded as a computer science term [6]_.
+
+:Map: Expresses the idea well, but has a lot of meanings in math and programming.
+
+:Use: Possibly confusing with C++'s ``using`` keyword; also not terribly descriptive.
+
+:Wrap: Passed in favor of "Adapt", which seems to be more common.
+
+
+Including "Type" in the name
+----------------------------
+
+Most of the terms being considered refer to general design patterns and, thus, using just the pattern's name may cause naming collisions or confusion as those design patterns may be used in other parts of the ROS codebase. 
+To reduce ambiguity, including the term selected with "Type" would make its usage clearer and help avoid name collisions;
+it should also make it easier for developers to find relevant documentation.
+
+
+Adding this feature in ``rclcpp``
+---------------------------------
+
+Placing this feature in ROS 2's C client library, ``rcl``, would allow this feature to be used in other client libraries, such as ``rclcpp`` and ``rclpy``.
+However, it is not clear that the difficulty of implementing this feature in ``rcl`` is worth the benefit to other client libraries.
+The primary client library that is expected to use this feature is ROS 2's C++ client library, ``rclcpp``.
+Placing this feature in ``rclcpp`` would allow implementation to take advantage of C++'s standard template library, and thus, speed up development.
```

Comment:
This paragraph is true in the sense that right now `rclcpp` has the `IntraProcessManager` - but I always like to point out that the more dedicated functionality we put in `rclcpp`, the more unlikely it is that any other language clients can be effectively used - or we are put into the situation of duplicating functionality across clients, slipping back in to one of `ros_comm`'s big problems.

So my request is "please can we not make this C++-specific functionality? Think of all the python CV people!"

Reply:
I agree with you, but in this case, with the intra process manager and the type adaption features, they are much safer and more efficient in the client library due to C++'s move semantics and ownership mechanics.

For intra process comms, it allows us to avoid type erasure (which would be necessary to place this functionality into `rcl`) and use ownership mechanics (unique and shared pointer) to ensure it is safely implemented. It could be implemented in `rcl`, but another added advantage of doing it in rclcpp, is that it reduce the number of function calls and calls that potentially are to functions in separate shared libraries.

In order to make use of type adaption in conjunction with the intra process communication we need to do it in rclcpp as well, at least partially.

We can still support a form of this feature in other languages using `rcl`/`rmw` in the future, but I believe it will require custom type support, and that might require that they are rmw specific. We're gonna explore this more, but I don't think we'll have time to implement it in Galactic. Where as the approach we're doing right now could be made in time for Galactic, and will have incremental benefits.

So it's a compromise, but we have considered it, and will continue to do so, even if we don't act on it right now.

Reply:
there are other languages where having type masquerading would be useful, e.g. rust (though there isn't an official client library for it)

Reply:
> there are other languages where having type masquerading would be useful, e.g. rust (though there isn't an official client library for it)

I didn't say it wouldn't be useful, just that the concrete benefits for C++ currently outweigh the potential benefits for existing or theoretical client libraries in other languages. 🤷‍♂️